### PR TITLE
#1821 nullpointer due to @BeanMapping via inheritance

### PR DESCRIPTION
**Pull request by sjaakd**
`mapExtended` does not have the `@BeanMapping`, but still it tries to resolve the prism. 

* I fixed this by copying the `SelectionParameters` for inheritance too and leaving out the `resultType`.
* Additionally I introduced a builder for `BeanMapping`.
* Also, I don't resort to the prims in `BeanMappingMethod`, but use `BeanMapping`

### CODE REVIEW

**Comment by filiphr** [comment]
@sjaakd what is the status of this one? I see that #1811 has been merged already

**Comment by sjaakd** [comment]
still an issue.. I know that @Captain1653 also looked at it.. Just need to find some time to think about this one.

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/BeanMapping.java [hunk 1]
```diff
@@ -36,87 +39,134 @@
...
-            beanMapping.ignoreUnmappedSourceProperties(),
-            builderMapping
-        );
+        public BeanMapping build() {
+
+            if ( prism == null ) {
+                return null;
+            }
+
+            Objects.requireNonNull( method );
```
[the remark this answers is not in the document]
Comment (sjaakd) [reply processor/src/main/java/org/mapstruct/ap/internal/model/source/BeanMapping.java]: you apply a fail fast and don't stumble upon the problem somewhere deep down in your code, where its often unclear which localvar / parameter is causing it.

Checkout the java doc of `Objects.requireNonNull`

**Comment by sjaakd** [comment]
> Do we really need a builder for the `BeanMapping`, why can't we keep the old methods? Looking at the builder it seems that it is the same as the methods before.

Yes.. but isn't that always the case..? For me I start to use builder when the number of constructor args starts to approach 10 args. Slowly we are moving all our classes to builders, so its unclear to me why we would apply that here.

btw: I added the mirror as constructor arg.

**Comment by sjaakd** [comment]
> Where exactly was the NPE before? There are a lot of changes to be able to determine what fixes the problem.

The problem was caused because we inherited the `resultType`. Check the selection parameters, it has a method `forInheritance` as well. Here the `resultType` is set to null.

I probably should have split stuff in different commits.

**Comment by filiphr** [comment]
We are passing the utils all over the place. Perhaps we should just pass the `ProcessorContext` everywhere and have a single parameter instead of 3 or 4, that is for some other day. I only mentioned it because it was difficult to review with so many changes 😄.

One think which is now not clear to me. If before we had an NPE, why isn't some test failing due to reporting?

**Comment by sjaakd** [comment]
> One think which is now not clear to me. If before we had an NPE, why isn't some test failing due to reporting?

I added a failing test :).

**Comment by filiphr** [comment]
Yes that test confused me. I thought that it was failing when we were reporting something. Anyways LGTM, feel free to merge

**Pull request merged**